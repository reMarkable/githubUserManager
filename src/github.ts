import { createAppAuth } from '@octokit/auth-app'
import { Octokit, type RestEndpointMethodTypes } from '@octokit/rest'

import * as mod from './github.js'
import { config } from './config.js'

export function getAuthenticatedOctokit(): Octokit {
  return new Octokit({
    authStrategy: createAppAuth,
    auth: {
      appId: config.githubAppID,
      privateKey: config.githubPrivateKey,
      installationId: config.githubInstallationID,
    },
  })
}

export async function getGithubUsersFromGithub(): Promise<Set<string>> {
  const octokit = mod.getAuthenticatedOctokit()
  const members = await octokit.paginate(octokit.orgs.listMembers, {
    org: config.githubOrg,
  })

  const pendingInvites = await octokit.paginate(octokit.orgs.listPendingInvitations, {
    org: config.githubOrg,
  })
  const pendingGithubAccounts = formatUserList(pendingInvites)

  const githubAccounts = formatUserList(members)

  if (pendingGithubAccounts.size > 0)
    console.log(`Outstanding GitHub invites for ${Array.from(pendingGithubAccounts).join(', ')}`)

  return new Set([...githubAccounts, ...pendingGithubAccounts])
}

export function formatUserList(users): Set<string> {
  return new Set(
    users
      .map((user) => user.login?.toLowerCase())
      .flat()
      .filter(Boolean),
  )
}

export async function getUserIdFromUsername(username: string): Promise<number> {
  const octokit = mod.getAuthenticatedOctokit()
  console.log(`Looking up user ${username}`)
  let user
  try {
    user = await octokit.users.getByUsername({ username })
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
  } catch (error) {
    throw `Unable to find user id for ${username}`
  }
  console.log(`User ${username} userid: ${user.data.id}`)
  return user.data.id
}

export async function addUsersToGitHubOrg(users: Set<string>): Promise<void> {
  for (const user of users) {
    await mod.addUserToGitHubOrg(user)
  }
}

export async function addUserToGitHubOrg(
  user: string,
): Promise<RestEndpointMethodTypes['orgs']['createInvitation']['response'] | boolean> {
  const octokit = mod.getAuthenticatedOctokit()
  if (config.ignoredUsers.includes(user.toLowerCase())) {
    console.log(`Ignoring add for ${user}`)
    return false
  }
  const userId = await mod.getUserIdFromUsername(user)
  console.log(`Inviting ${user} (${userId} to ${config.githubOrg})`)
  return await octokit.orgs.createInvitation({
    org: config.githubOrg,
    invitee_id: userId,
  })
}

export async function removeUsersFromGitHubOrg(users: Set<string>): Promise<void> {
  for (const user of users) {
    await mod.removeUserFromGitHubOrg(user)
  }
}

export async function removeUserFromGitHubOrg(
  user: string,
): Promise<RestEndpointMethodTypes['orgs']['removeMembershipForUser']['response'] | boolean> {
  const octokit = mod.getAuthenticatedOctokit()
  if (config.ignoredUsers.includes(user.toLowerCase())) {
    console.log(`Ignoring remove for ${user}`)
    return false
  }
  console.log(`Removing user/invitation ${user} from ${config.githubOrg}`)
  return octokit.orgs.removeMembershipForUser({
    org: config.githubOrg,
    username: user,
  })
}
