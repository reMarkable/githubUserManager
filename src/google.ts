import { google } from 'googleapis'

import * as mod from './google.js'
import { config } from './config.js'

export async function googleAuth() {
  const { googleCredentials, googleEmailAddress } = config
  const jwtClient = new google.auth.JWT(
    googleCredentials.client_email,
    null,
    googleCredentials.private_key,
    ['https://www.googleapis.com/auth/admin.directory.user.readonly'],
    googleEmailAddress,
  )
  await jwtClient.authorize()
  return jwtClient
}

export async function getAdminService() {
  return google.admin({
    version: 'directory_v1',
    auth: await googleAuth(),
  })
}

export async function getGithubUsersFromGoogle(): Promise<Set<string>> {
  const service = await mod.getAdminService()
  let githubAccounts = new Set<string>()
  let pageToken = null

  do {
    const userList = await service.users.list({
      customer: 'my_customer',
      maxResults: 250,
      projection: 'custom',
      fields: 'users(customSchemas/Accounts/github(value)),nextPageToken',
      customFieldMask: 'Accounts',
      pageToken: pageToken,
      query: config.removeSuspendedUsers ? 'isSuspended=false' : '',
    })
    pageToken = userList.data.nextPageToken
    githubAccounts = new Set([...githubAccounts, ...formatUserList(userList.data.users)])
  } while (pageToken != null)
  return githubAccounts
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function formatUserList(users: any[]): Set<string> {
  return new Set(
    users
      .map((user) =>
        user.customSchemas?.Accounts?.github?.map((account: { value: string }) => account.value?.toLowerCase().trim()),
      )
      .flat()
      .filter(Boolean),
  )
}
