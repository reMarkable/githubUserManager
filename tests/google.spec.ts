import { jest } from '@jest/globals'

jest.mock('googleapis')

import { google } from 'googleapis'
import * as mod from '../src/google'

const fakeUsersResponse = [
  { customSchemas: { Accounts: { github: [{ value: 'chrisns' }] } } },
  {
    customSchemas: {
      Accounts: { github: [{ value: 'Foo' }, , { value: 'tar' }] },
    },
  },
  {
    customSchemas: {
      Accounts: { github: [{ value: 'foo' }, { value: 'bar' }] },
    },
  },
]

describe('google integration', () => {
  beforeEach(() => {
    process.env.GOOGLE_EMAIL_ADDRESS = 'hello@example.com'
    process.env.GOOGLE_CREDENTIALS = Buffer.from(JSON.stringify({ client_email: 'foo', private_key: 'bar' })).toString(
      'base64',
    )
    jest.spyOn(global.console, 'log').mockImplementation(() => {})
  })
  it('googleAuth', () => {
    mod.googleAuth()
    return expect(google.auth.JWT).toMatchSnapshot()
  })
  it('getAdminService', () => {
    // FIXME: TypeError: google.admin.mockReturnValue is not a function
    // @ts-expect-error mock service isn't a complete implementation, so being lazy and just doing the bare minimum
    google.admin.mockReturnValue('adminservice')
    const result = mod.getAdminService()
    return expect(result).resolves.toBe('adminservice')
  })

  it('getGithubUsersFromGoogle', () => {
    const service = {
      users: {
        list: jest
          .fn<() => Promise<{ data: { users: typeof fakeUsersResponse } }>>()
          .mockResolvedValue({ data: { users: fakeUsersResponse } }),
      },
    }
    // FIXME: TypeError: Cannot assign to read only property 'getAdminService' of object '[object Module]'
    // @ts-expect-error mock service isn't a complete implementation, so being lazy and just doing the bare minimum
    jest.spyOn(mod, 'getAdminService').mockResolvedValue(service)

    const result = mod.getGithubUsersFromGoogle()
    expect(result).resolves.toMatchSnapshot()
  })

  it('formatUserList', () => expect(mod.formatUserList(fakeUsersResponse)).toMatchSnapshot())

  it('formatUserList bad', () =>
    expect(
      mod.formatUserList([
        {},
        { customSchemas: {} },
        { customSchemas: { Accounts: {} } },
        { customSchemas: { Accounts: { github: [] } } },
        { customSchemas: { Accounts: { github: [{}] } } },
        { customSchemas: { Accounts: { github: [{ value: 'chrisns' }] } } },
      ]),
    ).toMatchSnapshot())
})
