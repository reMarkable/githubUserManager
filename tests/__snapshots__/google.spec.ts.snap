// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`google integration formatUserList 1`] = `
Set {
  "chrisns",
  "foo",
  "tar",
  "bar",
}
`;

exports[`google integration formatUserList bad 1`] = `
Set {
  "chrisns",
  "userspace",
}
`;

exports[`google integration getGithubUsersFromGoogle 1`] = `
Set {
  "chrisns",
  "foo",
  "tar",
  "bar",
}
`;

exports[`google integration googleAuth 1`] = `
[MockFunction] {
  "calls": [
    [
      "foo",
      null,
      "bar",
      [
        "https://www.googleapis.com/auth/admin.directory.user.readonly",
      ],
      "hello@example.com",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
